File: Monotonic_Stack_29_Oct_2023/0402_remove_k_digits.py
class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        ## e.g Say k = 1
        ## 4321 : when digits are decreasing, pop from left, get 321
        ## 1234 : when numbers are increasing, pop from right, get 123
        s = []
        for n in num:
            while s and k and s[-1] > n:
                s.pop()
                k -= 1

            if s or n is not '0':
                s.append(n)

        if k:
            s = s[0:-k]

        return ''.join(s) or '0'

File: Monotonic_Stack_29_Oct_2023/0402-remove-k-digits.go
package main

func removeKdigits(num string, k int) string {
    s := make([]rune, 0)

    for _, r := range num {
        for top := len(s)-1; top >= 0 && k > 0 && s[top] > r; top-- {
            s = s[:top]
            k--
        }
        s = append(s, r)
    }

    if k > 0 {
        s = s[:len(s)-k]
    }

    for i := range s {
        if s[i] != '0' {
            return string(s[i:])
        }
    }

    return "0"
}
